src/main.ts

```diff
@@ -172,101 +108,145 @@ function getUnitHTML({ width, height }: AdViewManagerOptions, { unit, onLoadCode
-				const optsOverride = {
-					fallbackUnit: slot.fallbackUnit || opts.fallbackUnit,
-					minPerImpression: resMinPerImpression || opts.minPerImpression,
-					minTargetingScore: slot.minTargetingScore || opts.minTargetingScore,
-					targeting: slot.tags || opts.targeting,
-					acceptedReferrers: acceptedReferrers || opts.acceptedReferrers
-				}
-
-				this.options = { ...opts, ...optsOverride }
-				this.optsLoaded = true
+	private getTargetingInput(targetingInputBase: any, campaign: any): any {
+		const lastImpression = this.history.reverse().find(({ campaignId }) => campaignId === campaign.id)
+		return  {
+			...targetingInputBase,
+			'adView.navigatorLanguage': this.options.navigatorLanguage,
+			'adView.secondsSinceCampaignImpression': lastImpression
+				? Math.floor((Date.now() - lastImpression.time) / 1000)
+				: Infinity
 		}
 	}
+	private getStickyAdUnit(campaigns: any, acceptedReferrers: any): any {
+		if (this.options.disableSticky) return null
 
-	async getAdUnits(): Promise<any> {
-		const states = `status=${this.options.acceptedStates.join(',')}`
-		const publisherLimit = `limitForPublisher=${this.options.publisherAddr}`
-		const url = `${this.options.marketURL}/campaigns/with-targeting?${states}&${publisherLimit}`
-		const { campaigns, targeting } = await this.fetch(url).then(r => r.json())
-		this.options.targeting = this.options.targeting.concat(targeting)
-		return applySelection(campaigns, this.options)
+		const stickinessThreshold = Date.now() - IMPRESSION_STICKINESS_TIME
+		const stickyEntry = this.history.find(entry =>
+			entry.time > stickinessThreshold
+				&& entry.slotId === this.options.marketSlot
+		)
+		if (stickyEntry) {
+			const stickyCampaign = campaigns.find(campaign => campaign.id === stickyEntry.campaignId)
+			// Campaign couldn't be found, it means it's no longer active (either expired or unsound)
+			// in both cases, we don't want to be showing it
+			if (!stickyCampaign) return null
+			const { unit } = stickyCampaign.unitsWithPrice.find(x => x.unit.id === stickyEntry.unitId)
+			return {
+				unit,
+				price: '0',
+				acceptedReferrers,
+				html: getUnitHTML(this.options, { unit }),
+				isSticky: true
+			}
+		}
+		return null
 	}
-	async getFallbackUnit(): Promise<any> {
-		const { fallbackUnit } = this.options
-		if (!fallbackUnit) return null
-		const url = `${this.options.marketURL}/units/${this.options.fallbackUnit}`
-		const result = await this.fetch(url).then(r => r.json())
-		return result.unit
+	private isCampaignSticky(campaign: any): boolean {
+		if (this.options.disableSticky) return false
+		const stickinessThreshold = Date.now() - IMPRESSION_STICKINESS_TIME
+		return !!this.history.find(entry => entry.time > stickinessThreshold && entry.campaignId === campaign.id)
+	}
+	async getMarketDemandResp(): Promise<any> {
+		const marketURL = this.options.marketURL
+		const depositAsset = this.options.whitelistedTokens.map(tokenAddr => `&depositAsset=${tokenAddr}`).join('')
```

Comment:
Will this work? IIRC if we want a list (array) in the url `depositAsset[]=` should be used, otherwise I think it will get the last mentioned in the URL `&depositAsset=`

I know we currently support both (string & array) but if we add another whitelisted Token we might forget this place

Reply:
yes, cause it chains them with & so it will result in an array on the server